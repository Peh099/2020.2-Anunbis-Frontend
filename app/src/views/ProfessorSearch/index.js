import React, { useState } from 'react';
import { useParams } from 'react-router-dom';
import Feed from '../../components/Feed';
import api from '../../services/Api';
import Loading from '../../components/Loading';
import {
  Container,
  FoundDiv,
  Img,
  FoundHeader,
  Name,
  Discipline,
  LoadingBox,
  NotFound,
} from './styles';
import BtnOptions from '../../assets/images/Btn_options.png';
import Users from '../../services/Users';

const ProfessorFound = ({
  professor,
  onClick,
  isSelected,
  setSelectedDiscipline,
}) => {
  const [showDisciplines, setShowDisciplines] = useState(false);

  function onClickDiscipline(index) {
    onClick();
    setSelectedDiscipline(index);
  }

  return (
    <FoundDiv>
      <FoundHeader selected={isSelected}>
        <Img
          src={BtnOptions}
          onClick={() => setShowDisciplines(!showDisciplines)}
          rotate={showDisciplines ? '90deg' : ''}
          data-testid="img-1"
        />
        <Name onClick={onClick}>{professor.name}</Name>
      </FoundHeader>

      {showDisciplines &&
        professor.disciplines.map((disci, index) => (
          <Discipline
            onClick={() => onClickDiscipline(index)}
            key={disci.discipline_code}
          >
            {`[${disci.discipline_code}] ${disci.name}`}
          </Discipline>
        ))}
    </FoundDiv>
  );
};

function getPostsByDiscipline(professor, disciplineSelected) {
  const { discipline_code } = professor.disciplines[disciplineSelected]; // eslint-disable-line
  return professor.posts.filter(
    (post) => post?.discipline.discipline_code === discipline_code, // eslint-disable-line
  );
}

function getPosts(professor, disciplineSelected) {
  if (professor && professor.posts) {
    return disciplineSelected < 0
      ? professor.posts
      : getPostsByDiscipline(professor, disciplineSelected);
  }
  return [];
}

function getFeedbacks(professor, posts, disciplineSelected) {
  if (disciplineSelected < 0) return professor;
  if (posts.length === 0) return -1;

  const rates = {
    rating: professor.rating,
    didactic: professor.didactic,
    metod: professor.metod,
    avaliations: professor.avaliations,
    disponibility: professor.disponibility,
  };

  return rates;
}

function ProfessorSearch() {
  const { professorName } = useParams();
  const [professors, setProfessors] = useState([]);
  const [selected, setSelected] = useState({
    professor: 0,
    discipline: -1,
  });
  const [newAvaliationState, setNewAvaliationState] = useState(false);
  const professor = professors[selected.professor];
  const posts = getPosts(professor, selected.discipline);
  const feedbacks = getFeedbacks(professor, posts, selected.discipline);
  const [loading, setLoading] = useState(true);

  function handleSetSelected(indexProfessor, indexDiscipline) {
    setLoading(true);
    setSelected({
      professor: indexProfessor,
      discipline: indexDiscipline,
    });
  }

  React.useEffect(() => {
    if (professors.length === 0) return;
    setLoading(true);
    const { id_professor } = professors[selected.professor]; // eslint-disable-line
    const startRequest = new Date().getTime();

    api.get(`/professor/${id_professor}`).then((response) => {
      // eslint-disable-line
      const requestDuration = startRequest - new Date().getTime();
      if (response.status === 200) {
        professors[selected.professor] = response.data;
        setProfessors(professors);
      }
      setTimeout(
        () => {
          setLoading(false);
        },
        requestDuration > 500 ? 0 : 500 - requestDuration,
      );
    });
  }, [professors, newAvaliationState, selected]);

  React.useEffect(() => {
    api
      .get(`/professor/${professorName}`)
      .then((response) => {
        if (response.status === 200) {
          setProfessors(response.data);
          handleSetSelected(0, -1);
        }
        if (response.data.length === 0) setLoading(false);
      })
      .catch(() => {
        setProfessors([]);
      });
  }, [professorName]);

  return (
    <Container hasProfessors={professors.length > 0}>
      {professors.length > 0 && (
        <Feed
          title="Professores"
          width="min(20vw, 210px)"
          radius="0px 0px 10px 10px"
          key={professors.length}
        >
          {professors.map((prof, index) => (
            <ProfessorFound
              professor={prof}
              onClick={() => handleSetSelected(index, -1)}
              setSelectedDiscipline={(indexDiscipline) =>
                handleSetSelected(index, indexDiscipline)
              }
              key={prof.id_professor}
              isSelected={index === selected.professor}
            />
          ))}
        </Feed>
      )}
      <Feed
        title={professor ? `${professor.name}` : 'Pesquisa de professor'}
        radius="0px 0px 10px 10px"
      >
        {professors.length > 0 && (
          <Feed.Header
            professor={professor}
            feedbacks={feedbacks}
            canAvaliate={Users.STUDENT.isAuthenticated()}
            onNewAvaliation={() => setNewAvaliationState(!newAvaliationState)}
            backColor="#FFFDE7"
            border="1px solid var(--black)"
          />
        )}
        {professors.length === 0 && !loading && (
          <NotFound>
            <div>Nenhum Professor Encontrado!</div>
          </NotFound>
        )}
        {professors.length > 0 && (
          <Feed.Title zIndex="0">
            {posts.length === 0 && !loading
              ? 'Sem Avaliações Ainda'
              : 'Avaliações'}
          </Feed.Title>
        )}
        {!loading && <Feed.PostsBox posts={posts} key={posts.length} />}
        {loading && (
          <LoadingBox data-testid="load-1">
            <Loading />
          </LoadingBox>
        )}
      </Feed>
    </Container>
  );
}

export default ProfessorSearch;
