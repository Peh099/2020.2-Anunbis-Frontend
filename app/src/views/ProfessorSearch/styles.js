import styled from 'styled-components';

export const Container = styled.div`
  display: ${(props) => (props.hasProfessors ? 'grid' : 'flex')};
  grid-template-columns: min(25vw, 240px) min(80vw, 700px);
  flex-direction: row;
  @media (max-width: 760px) {
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
    width: 90vw;
    // grid-template-columns: repeat(auto-fill, minmax(250px, 80vw));
  }
`;

export const FoundDiv = styled.div`
  margin: 10px;
  background-color: var(--yellow);
  border-radius: 5px;
  box-shadow: 1px 1px rgba(0, 0, 0, 50%);
  padding: 5px;

  * {
    cursor: pointer;
  }
`;

export const FoundHeader = styled.div`
  display: flex;
  flex-direction: row;
  opacity: ${(props) => (props.selected ? '100%' : '60%')};
  font-weight: ${(props) => (props.selected ? 'bold' : 'normal')};
`;

export const Img = styled.img`
  height: 20px;
  transition: transform 0.5s ease;
  transform: ${(props) => `rotate(${props?.rotate})`};
`;

export const Name = styled.div`
  display: inline-block;
  width: 100%;
  @media (max-width: 760px) {
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
  }
`;

export const Discipline = styled.div`
  padding: 5px;
  margin: 5px;
  background-color: #fff9c4;
  border-radius: 10px;
  box-shadow: 2px 2px rgba(0, 0, 0, 50%);
  text-align: center;
  @media (max-width: 760px) {
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
  }
`;

export const LoadingBox = styled.div`
  width: 100%;
  height: 20vh;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const NotFound = styled.div`
  height: 20vh;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  font-family: 'Comfortaa', cursive;
  font-weight: bold;
`;
