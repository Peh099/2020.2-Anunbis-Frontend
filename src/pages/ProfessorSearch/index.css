.ProfessorSearch {
        width: 45%;
        height: 90%;
}