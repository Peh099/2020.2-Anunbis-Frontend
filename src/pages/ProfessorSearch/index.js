import React, { useState } from 'react';
import {useParams} from "react-router-dom";

const ProfessorItem = ({professor}) =>{
    if(professor.length === 0){
        return (<div>Nenhum professor encontrado</div>)
    }
    return (<div>
        {professor.name}
    </div>)
}


export default function ProfessorSearch(props) {
    const {professorName} = useParams();
    const [professors, setProfessors ]= useState([]);

    React.useEffect(async ()=>{
        const url = process.env.REACT_APP_API_HOST + "/professor/" + professorName;
        const response = await fetch(url);
        const data = await response.json();
        if(response.ok){
            setProfessors(data);
        }
    },[]);

    return (<div>
            {professors.map(prof => <ProfessorItem professor={prof}></ProfessorItem>)}
          </div>);
}