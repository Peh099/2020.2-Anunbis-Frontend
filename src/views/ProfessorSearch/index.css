.ProfessorSearch {
        width: 45%;
        height: 90%;
        display: flex;
        align-items: center;
}

.ProfessorSearch .Professors {
        width: 100%;
        height: 100%;
        display: flex;
        flex-direction: column;
        align-items: center;
}

.ProfessorSearch .avaliationProfBox{
        position: absolute;
        width: 450px;
        height: 300px;
        display: flex;
        align-items: center;
        justify-content: center;
}