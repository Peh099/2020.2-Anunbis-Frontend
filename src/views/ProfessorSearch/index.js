import React, { useState } from 'react';
import styled from 'styled-components';
import { useParams } from "react-router-dom";
import Feed from '../../components/Feed';
import ProfessorBox from '../../components/ProfessorBox';
import Avaliation from '../../components/Avaliation';

const Content = styled.div`
display: flex;
align-items: center;
`
const AvaliationProfBox = styled.div`
        position: absolute;
        width: 450px;
        display: flex;
        align-items: center;
        justify-content: center;
`
;



export default function ProfessorSearch() {
    const { professorName } = useParams();
    const [professors, setProfessors] = useState([]);
    const [boxAvaliation, setBoxAvaliation] = React.useState('');

    React.useEffect(() => {
        async function fetchData() {
            const url = process.env.REACT_APP_API_HOST + "/professor/" + professorName;
            const response = await fetch(url);
            const data = await response.json();
            if (response.ok) {
                setProfessors(data);
            }
        }
        fetchData();
    }, [professorName]);



    function makeAvaliation(id_professor, name, disciplines) {
        setBoxAvaliation(
            <Avaliation
                close={() => setBoxAvaliation('')}
                reg_student={JSON.parse(localStorage.getItem('student')).reg_student}
                id_professor={id_professor}
                name_professor={name}
                disciplines={disciplines}
            />
        )
    }

    const Professors = (({ professors }) => {
        const Conteiner = styled.div`
        height: 400px;
        display: flex;
        flex-direction: column;
        align-items: center;
        justify-content: flex-start;
        `;


         return (
            
                <Feed title={professorName}>
                    <Conteiner>
                {professors?.map(prof => {
                    return (
                        <ProfessorBox onClick={() => makeAvaliation(prof.id_professor, prof.name, prof.disciplines)} name={prof.name} rating={prof.rating} posts={prof.posts}></ProfessorBox>
                    )
                })}
                </Conteiner>
                </Feed>
            
         );
    });

    return (
        <Content>
            <AvaliationProfBox>
                {boxAvaliation}
            </AvaliationProfBox>
                <Professors professors={professors}/>
        </Content>
    );
}