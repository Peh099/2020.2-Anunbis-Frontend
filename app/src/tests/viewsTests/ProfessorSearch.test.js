import renderer from 'react-test-renderer';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import ProfessorSearch from '../../views/ProfessorSearch';

describe('Snapshot ProfessorSearch with Posts', () => {
  it('matches the snapshot', () => {
    const history = createMemoryHistory();
    history.push('/user/professor/search/:professorName');

    const tree = renderer
      .create(
        <Router history={history}>
          <ProfessorSearch history={history} />
        </Router>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
