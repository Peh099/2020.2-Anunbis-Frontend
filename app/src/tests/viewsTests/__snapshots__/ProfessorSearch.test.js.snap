// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot ProfessorSearch with Posts matches the snapshot 1`] = `
<div
  className="sc-hiKfDv cwXlMa"
>
  <div
    className="sc-bdnxRM LfQdx"
    radius="0px 0px 10px 10px"
    title="Sem Resultados"
  >
    <div
      className="sc-gtsrHT iUwlZJ"
    >
      Sem Resultados
    </div>
    <div
      className="sc-jSFjdj cGLgru"
    >
      <div>
        Nenhum Professor Encontrado!
      </div>
    </div>
  </div>
</div>
`;
